crates/exex/exex/src/backfill/stream.rs

```diff
@@ -11,90 +11,159 @@ use futures::{
+    fn poll_next_task(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<T, BlockExecutionError>>> {
+        match ready!(self.tasks.poll_next_unpin(cx)) {
+            Some(res) => Poll::Ready(Some(res.map_err(|e| BlockExecutionError::Other(e.into()))?)),
+            None => Poll::Ready(None),
+        }
     }
 }
 
-impl<E, P> Stream for BackFillJobStream<E, P>
+impl<E, P> Stream for StreamBackfillJob<E, P, SingleBlockStreamItem>
 where
     E: BlockExecutorProvider + Clone + Send + 'static,
     P: HeaderProvider + BlockReader + StateProviderFactory + Clone + Send + 'static + Unpin,
 {
-    type Item = Result<(BlockWithSenders, BlockExecutionOutput<Receipt>), BlockExecutionError>;
+    type Item = Result<SingleBlockStreamItem, BlockExecutionError>;
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         let this = self.get_mut();
 
         // Spawn new tasks only if we are below the parallelism configured.
         while this.tasks.len() < this.parallelism {
+            // If we have a block number, then we can spawn a new task for that block
             if let Some(block_number) = this.range.next() {
-                let task = this.spawn_task(block_number);
+                let mut job = SingleBlockBackfillJob {
+                    executor: this.executor.clone(),
+                    provider: this.provider.clone(),
+                    range: block_number..=block_number,
+                    stream_parallelism: this.parallelism,
+                };
+                let task =
+                    tokio::task::spawn_blocking(move || job.next().expect("non-empty range"));
                 this.tasks.push_back(task);
             } else {
                 break;
             }
         }
 
-        match ready!(this.tasks.poll_next_unpin(cx)) {
-            Some(res) => Poll::Ready(Some(res.map_err(|e| BlockExecutionError::Other(e.into()))?)),
-            None => Poll::Ready(None),
+        this.poll_next_task(cx)
+    }
+}
+
+impl<E, P> Stream for StreamBackfillJob<E, P, BatchBlockStreamItem>
+where
+    E: BlockExecutorProvider + Clone + Send + 'static,
+    P: HeaderProvider + BlockReader + StateProviderFactory + Clone + Send + 'static + Unpin,
+{
+    type Item = Result<BatchBlockStreamItem, BlockExecutionError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        // Spawn new tasks only if we are below the parallelism configured.
+        while this.tasks.len() < this.parallelism {
+            // Take the next `batch_size` blocks from the range and calculate the range bounds
+            let mut range = this.range.by_ref().take(this.batch_size);
+            let start = range.next();
+            let range_bounds = start.zip(range.last().or(start));
+
+            // Advance the range by `batch_size` blocks
+            this.range.nth(this.batch_size);
```

Comment:
this will run a few times but should be negligible overhead but range could be wrapped into Peekable and integrated into while condition

but it looks like this is just 

https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.chunks

Reply:
using itertools' chunks will require having a lifetime on the struct and also storing the original iterator in a field. We can also do `(0..batch_size).map_while(|_| this.range.next())` here, but it's more expensive than doing `nth` just once.

Didn't get your idea with peekable.